### PR TITLE
Outline bridged Objective C methods

### PR DESCRIPTION
**Pull request by aschwaighofer**
Add a SIL Outliner pass that outlines the bridging of Objective C calls.

rdar://33387700

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

File: lib/SILOptimizer/Transforms/Outliner.cpp [hunk 1]
```diff
@@ -0,0 +1,1306 @@
...
+///    %41 = enum $Optional<String>, #Optional.some!enumelt.1, %40 : $String
+///    br bb10(%41 : $Optional<String>)
+///
+///  bb9:
+///    %43 = enum $Optional<String>, #Optional.none!enumelt
+///    br bb10(%43 : $Optional<String>)
+///
+///  bb10(%45 : $Optional<String>):
+class BridgedProperty : public OutlinePattern {
+  std::string outlinedName;
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/Outliner.cpp]: Question, why capitalize this field differently?

File: lib/SILOptimizer/Transforms/Outliner.cpp [hunk 2]
```diff
@@ -0,0 +1,1306 @@
...
+    auto *Arg = OutlinedEntryBB->getArgument(0);
+    FirstInst->setOperand(0, Arg);
+    PropApply->setArgument(0, Arg);
+  }
+  Builder.setInsertionPoint(OldMergeBB);
+  Builder.createReturn(ClassMethod->getLoc(), OldMergeBB->getArgument(0));
+  return std::make_pair(Fun, std::prev(StartBB->end()));
+}
+
+#define ADVANCE_ITERATOR_OR_RETURN_FALSE(It)                                   \
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/Outliner.cpp]: There is something in STLExtras similar to this called next_or_end. Just an FYI. But I guess you are using the macro so you can return easier?

Regardless, shouldn't this be wrapped in a:

```
do { ... } while (0);
```

File: lib/SILOptimizer/Transforms/Outliner.cpp [hunk 3]
```diff
@@ -0,0 +1,1306 @@
...
+  auto *Load = dyn_cast<LoadInst>(It);
+
+  // Otherwise, trying matching from the method call.
+  if (!Load) {
+    // Try to match without the load/strong_retain prefix.
+    auto *CMI = dyn_cast<ClassMethodInst>(It);
+    if (!CMI)
+      return false;
+    FirstInst = CMI;
+  } else FirstInst = Load;
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/Outliner.cpp]: The way this statement is at the end of the line after the else seems weird... did you intend for this? Or did git-clang-format do this?

File: lib/SILOptimizer/Transforms/Outliner.cpp [hunk 4]
```diff
@@ -0,0 +1,1306 @@
...
+      return;
+
+    // Dump function if requested.
+    if (DumpFuncsBeforeOutliner.size() &&
+        Fun->getName().find(DumpFuncsBeforeOutliner, 0) != StringRef::npos) {
+      Fun->dump();
+    }
+
+    SmallVector<SILFunction *, 16> FunctionsAdded;
+		bool Changed = tryOutline(Fun, FunctionsAdded);
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/Outliner.cpp]: The indenting here is off.

File: lib/SILOptimizer/Transforms/Outliner.cpp [hunk 5]
```diff
@@ -0,0 +1,1306 @@
...
+       } else if (isa<TermInst>(CurInst)) {
+         std::copy(CurBlock->succ_begin(), CurBlock->succ_end(),
+                   std::back_inserter(Worklist));
+         ++CurInst;
+       } else {
+         ++CurInst;
+       }
+    }
+  }
+	return false;
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/Outliner.cpp]: The indentation here is off.

**Review by gottesmm — COMMENTED** [review]
Just read the code for my own understanding of what we are doing. Saw a few small nits.

**Comment by jrose-apple** [comment]
Sounds like we're going to need a form of benchmark that does code size comparisons.

**Comment by gottesmm** [comment]
@jrose-apple 
4aa725edb5f1df40042b145c0b04a0a93ab62371d308c49ee7072f34dc1693ac

**Comment by aschwaighofer** [comment]
I use the source compatibility suite to look at code size. An automated way at getting the numbers from our CI system would be ideal.

**Comment by eeckstein** [comment]
Nice work!

File: lib/SILOptimizer/Transforms/Outliner.cpp (hunk 1 above)
Comment (aschwaighofer) [reply lib/SILOptimizer/Transforms/Outliner.cpp]: Too much time in IRGen ...

File: lib/SILOptimizer/Transforms/Outliner.cpp (hunk 2 above)
Comment (aschwaighofer) [reply lib/SILOptimizer/Transforms/Outliner.cpp]: I use this macro to make the return on end() easier.

File: lib/SILOptimizer/Transforms/Outliner.cpp (hunk 3 above)
Comment (aschwaighofer) [reply lib/SILOptimizer/Transforms/Outliner.cpp]: No clang-format puts it on the next line.

File: lib/SILOptimizer/Transforms/Outliner.cpp (hunk 5 above)
Comment (aschwaighofer) [reply lib/SILOptimizer/Transforms/Outliner.cpp]: Somehow tabs made it into this file ...

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Why is this a separate outlining pass and not just done as part of SILGen?

**Comment by aschwaighofer** [comment]
This only runs under -Osize and matching this from the AST during SILGen did not seem the right place to me.

**Comment by slavapestov** [comment]
@aschwaighofer I think it makes more sense to have SILGen always outline these operations actually. While your approach is fine for now, in the long term we might want to rework SILGen instead, and have the performance inliner inline the outlined thunks when not running in -Osize.

**Pull request merged**